Fix conditions that prevent from editing or removing the site's owner

wojtekn opened the pull request:
#### Proposed Changes

Currently, there are conditions that prevent from displaying UX that allows editing or removing the site's owner. Those do not include Atomic sites, and this PR modifies those to include Atomic sites too.

This was tracked under https://github.com/Automattic/wp-calypso/issues/3140  and was done for simple sites under https://github.com/Automattic/wp-calypso/pull/57042 This PR should fix the problem that remains for Atomic sites. Behavior for non-Atomic Jetpack sites (aka self-hosted) is not changed.

This PR addresses the UX issue - it removes UX that shouldn't be displayed at the first place. There is also D84810-code that addresses the server-side part of this problem - fixes checking if editing or removing the owner is allowed in such cases.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

##### Atomic sites

Steps to reproduce:

1. Create an Atomic site
2. Navigate to WPCOM -> Users
3. Invite a new user as Administrator - another WPCOM account
4. Accept invitation received by email
5. Logged as an invited user, navigate to WPCOM -> Users
6. Click on the site's owner

I shouldn’t see the UI that allows for editing the owner’s role or deleting it, and I should see the message instead:

![Screen Shot 2022-07-27 at 13 31 10](https://user-images.githubusercontent.com/727413/181709699-2f8be43d-e066-4b33-a23e-44023ab1fdf6.png)


##### Self-hosted Jetpack sites

Steps to reproduce:

1. Create an ephemeral site (or any site that is initially not connected to WPCOM)
2. Navigate to Plugins
3. Click “Set up Jetpack” and continue the process (it can be Jetpack Free)
4. Navigate to WPCOM -> Users
5. Invite a new user as Administrator - another WPCOM account
6. Accept invitation received by email
7. Logged as an invited user, navigate to WPCOM -> Users
8. Click on the site's owner

I should still see the UI that allows for editing the owner’s role or deleting it:

![Screen Shot 2022-07-29 at 09 39 53](https://user-images.githubusercontent.com/727413/181709678-b6abb7f0-882d-4dd8-a72a-8d6ecbfedee5.png)


#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] ~~[Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?~~
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] ~~Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)~~
- [ ] ~~Have you sent any new strings for translation (PCYsg-1vr-p2) ASAP?~~

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to p9F6qB-9M6-p2

wojtekn committed:
Fix conditions that prevent from editing or removing the site's owner

Currently, there are conditions that prevent from displaying UX
that allows editing or removing the site's owner. Those do not include
Jetpack sites, and this PR modifies those to include Jetpack sites too.

vykes-mac reviewed (commented):
@wojtekn This works from Calypso but I'm still able to edit/delete user from wp-admin/users side of things. It seems wp-admin calls out to a different endpoint than calypso. Might want to apply this fix on that side of things also.

danielbachhuber commented on client/my-sites/people/edit-team-member-form/edit-user-form.jsx:
```diff
@@ -98,10 +98,14 @@ class EditUserForm extends Component {
+			( ( ! isJetpack && user.ID === siteOwner ) ||
+				( isJetpack && user.linked_user_ID === siteOwner ) ) &&
+			user.ID !== currentUser.ID
+		) {
```
It would be great to encapsulate this logic to a function, and write some tests around the function.

Given there are multiple scenarios covered by this conditional, tests would explicitly document the behavior for each scenario.

wojtekn replied:
Having the server-side tests, I'm thinking if we can skip writing tests for Calypso UX at this point. Currently, the conditions are documented via docblock comments and are written similarly to other conditions in this component.

Another thing is that we don't have any tests for this form yet. If we want to write them, we need to a) extract the current condition to separate component to avoid duplication and then write tests or b) write tests on the component level (mock data, render, check what it renders). I'm not sure if the effort would be reasonable now.

danielbachhuber replied:
@wojtekn I'm fine with this assessment.

danielbachhuber commented on client/my-sites/people/delete-user/index.jsx:
```diff
@@ -224,7 +224,10 @@ class DeleteUser extends Component {
 		const { translate, isJetpack, siteOwner, user } = this.props;
 
 		// A user should not be able to remove the site owner.
-		if ( ! isJetpack && user.ID === siteOwner ) {
```
I think it probably makes sense to have Jetpack continue to mirror core behavior because there isn't a clear concept of 'site owner' for a Jetpack-connnected site.

wojtekn replied:
Thanks, I updated PR and aligned the logic with WP Admin.

wojtekn committed:
Modify code that prevents from deleting owner to not catch self-hosted Jetpack

Admin user can delete other admin users on the self-hosted Jetpack sites.
I modify the logic that prevents from removing owner to catch only Atomic
and non-Jetpack sites.

sejas reviewed (approved):
Tested and working as described. 👍
